lib/src/bound_multipart_stream.dart

```diff
@@ -133,7 +133,11 @@ class BoundMultipartStream {
             _subscription.pause();
             _buffer = data;
             _index = 0;
-            _parse();
+            try {
+              _parse();
+            } catch (e, st) {
+              _controller.addError(e, st);
```

Comment:
I don't think it's safe to continue parsing after an error here.
The `_parse` function bailed out in the middle of its input, so it's impossible to say whether there is some multipart delimiter later in the current input, and just emitting the error and continuing to the next data block seems unsafe.

Maybe just do `_subscription.pause()`. That prevents any further data, but still requires the listener to cancel the stream (which will provide any error from `_subscription.cancel()` to them).

(This stream should really be propagating the future returned by `_subscription.cancel` as the return-value of its own `onCancel` callback, it currently doesn't).

Reply:
Thanks - you're right, it looks like it is possible to keep pumping data into this transformer if parsing fails. I'll add another test to capture this. It also looks like I need to cleanup the `_multipartController`, if there currently is one.

I am not sure what to do with the input stream if parsing fails. I'm leaning towards just cancelling the subscription on parse error, and continuing to propagate the error to the output stream (`_controller.stream`). I am confused that this error is not already being propagated - my assumption was that an error thrown in the subscription's `onData` callback would trigger its `onError` callback.

Reply:
> my assumption was that an error thrown in the subscription's onData callback would trigger its onError callback.

That's actually not the case. Events are generally treated as top-level/microtask event-loop triggers with no context, so the `onData` handler is fire-and-forget. It gets the value, and anything after that is its own responsibility. If it throws, the error will become an uncaught error in the current zone. The only one who can send events on a stream is the one who created it, the `onDone` handler cannot do that.

You might be thinking of the behavior of transformations like `Stream.map`, where an error in the converter function becomes an error on the output stream, but that's because the operation converts one stream to another, and the code running the convert function is the code putting data and errors into the output stream.

Reply:
I would definitely cancel the source subscription on a parse error, then send an error event on the output stream, followed by a done event.

This should all happen synchronously in response to the input package that caused the error, so you won't have seen any other events from the source stream yet.